src/lib.rs

```diff
@@ -1,7 +1,317 @@
+    fn get_help(&self) -> &Manual;
+}
+
+#[derive(Debug)]
+pub struct Manual<'a> {
+    short_description: &'a str,
+    detailed_help: Vec<&'a str>,
+}
+
+impl<'a> Manual<'a> {
+    pub fn new(short_description: &'a str, detailed_help: Vec<&'a str>) -> Manual<'a> {
+        Manual {
+            short_description,
+            detailed_help,
+        }
+    }
+}
+
+pub struct Runtime {}
+
+pub struct Parser<'a> {
+    verbs: HashMap<String, Verb<'a>>,
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(verbs: Vec<Verb>) -> Parser {
+        let mut verbs_map = HashMap::with_capacity(verbs.len());
+
+        for verb in verbs {
+            verbs_map.insert(verb.name().clone(), verb);
+        }
+
+        Parser {
+            verbs: verbs_map,
+        }
+    }
+
+    pub fn parse(&self, input: String) -> Action {
+        let (verb_name, remaining_commands, matching_verb) = match input.split_once(" ") {
+            Some((verb_name, remaining_commands)) => {
+                (verb_name.to_owned(), remaining_commands, self.verbs.get(verb_name))
+            },
+            None => {
+                (input.clone(), "", self.verbs.get(&input))
+            },
+        };
+
+        let action = match matching_verb {
+            Some(verb) => {
+                verb.parse(remaining_commands)
+            },
+            None => Action::Unknown(verb_name),
+        };
+
+        action
+    }
+}
+
+#[derive(Debug)]
+pub struct Verb<'a> {
+    name: String,
+    verbs: HashMap<String, Verb<'a>>,
+    commands: HashMap<String, Command>,
+    manual: Manual<'a>,
+}
+
+impl<'a> Verb<'a> {
+    pub fn new(name: &str, verbs: Option<Vec<Verb<'a>>>, commands: Option<Vec<Command>>, manual: Manual<'a>) -> Verb<'a> {
+        let mut verbs_map = HashMap::new();
+
+        if let Some(verbs) = verbs {
+            for verb in verbs {
+                verbs_map.insert(verb.name().clone(), verb);
+            }
+        }
+
+        let mut commands_map = HashMap::new();
+
+        if let Some(commands) = commands {
+            // TODO: Need to check if a verb with the same name already exists.
+            for command in commands {
+                commands_map.insert(command.name().clone(), command);
+            }
+        }
+
+        Verb {
+            name: name.to_owned(),
+            verbs: verbs_map,
+            commands: commands_map,
+            manual,
+        }
+    }
+
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    pub fn parse(&self, input: &str) -> Action {
+        let (command_name, remaining_commands) = match input.split_once(" ") {
+            Some((command_name, remaining_commands)) => (command_name.to_owned(), remaining_commands),
+            None => (input.to_owned(), ""),
+        };
+
+        if self.verbs.contains_key(&command_name) {
+            let command_verb = self.verbs.get(&command_name).expect(format!("Expected there would be a verb named '{}' but couldn't find it.", command_name).as_str());
```

Comment:
Generally you should prefer `unwrap_or_else` and pass in a closure instead of `expect` when you're making a function call, to defer execution of the unhappy path - rather than always invoking the unhappy path as you do here (even though it is just a format / as_str call).